aip/general/0180.md

```diff
@@ -139,6 +139,84 @@ code, as such changes will be seen as breaking by those users.
 whether a proposed change is likely to break users, and an expansive reading of
 this guidance could ostensibly prevent _any_ change (which is not the intent).
 
+#### Default values must not change
+
+Changing the default value is considered breaking and **must not** be done. The
+default behavior for a resource is determined by it's default values, and this
+**must not** change across minor versions.
+
+For example:
+
+```proto
+// A representation of an automobile
+message Automobile {
```

Comment:
Side note: this PR feels like it's mostly thinking about clients creating new resources... but what about default values populated when new fields are introduced?

Suppose Automobile originally didn't have a `wheels` field at all, and I create an automobile with name X.

If I fetch X later, after the server has been updated to introduce `wheels`, with a later wheels-aware client, should the field:

- Not be populated
- Always populated with the value that it would have if I created a new automobile now without specifying the number of wheels
- Potentially be given a value *different* to the one that would be created for a new automobile

The ability to have different "migrated" vs "newly created" defaults could be useful, I suspect - would it cause Terraform issues?

Reply:
Default values for new fields that are introduced are the cause of the types of bugs in https://github.com/hashicorp/terraform-provider-google/issues/14264#issuecomment-1504273202 and https://github.com/hashicorp/terraform-provider-google/issues/14203#issuecomment-1496654588

The cases that cause issues for Terraform are when a field has different behavior depending on when the resource was provisioned.

Often this happens when a Terraform provider developer adds support for a new field and assumes that the field is always populated in the API response due to a server-set default value. This may be true for all instances of the resource created after the field was introduced, but the field may not be returned for older resources. The developer typically doesn't have access to older resources and so can't tell if that's the case until a public release happens and customers run into issues.

To me this is pretty clearly a failure of the API to have a consistent experience. Clients shouldn't need to be aware of different types of behavior for a field depending on the date the resource was provisioned. Having consistent behavior on the field by always returning the server-set value even if the default for that value may have changed is important.

Reply:
> Clients shouldn't need to be aware of different types of behavior for a field depending on the date the resource was provisioned.

I'm not so sure about that. I agree with it as a general rule, but I suspect there are plenty of cases where it's actually really useful. The example that comes to mind is around encryption:

- You don't want an encryption type for a single resource to vary over time
- It *is* useful to be able to say "If you don't specify an encryption type when creating the resource, we'll use the strong currently-available one"

(The same might be true of some other quality metrics - audio/video/image encoding, for example. Today's default might be effectively obsolete in 10 years... it would be really unfortunate to have to keep using it as the default when much better options are available.)

I can absolutely see that causes a problem for declarative clients - but I think it's worth acknowledging as a scenario where there are conflicting benefits.

> Having consistent behavior on the field by always returning the server-set value even if the default for that value may have changed is important.

Definitely agree on this one.

Reply:
I think that we do need to have static defaults at the API level. Different clients have different assumptions and tolerance for changes in defaults, and what works for a client like gcloud may not work for declarative clients.

As a broad assumption we should make it true that the same API request creates the same resource on an API regardless of the time when that request is made. Things like new features will happen, but the functionality of the resource shouldn't be impacted by changes to default values.

I propose that changes to default values should be made at the client level rather than globally at the API level. This can be accomplished by the API freezing behavior on unset values, while clients can update the default value sent as appropriate for their needs.

In practice this would involve clients like Console updating the default values fairly often, gcloud maybe a bit less often, and declarative clients the least often (though it should happen at some point!). We may want additional information on "recommended" defaults to be published by the API somewhere, so that clients can identify changes and make them on their own time.

Reply:
We definitely need static defaults in the API, at least for the case that started this thread, which is new fields with default values, like wheels=4. Whether/when/where default values should be returned by the API is another issue. Kubernetes always returns the default values, but it also deals with merging in server-side apply and strategic merge patch.

Changing a static default is a non-backward-compatible API change. That's the position we took in Kubernetes. If an API plans on changing such values, they become dynamic defaults, essentially, and the service needs to persist the selected value and return it somehow/somewhere.

Users of declarative clients generally expect the effective desired state to be deterministic, so changing a default in such a client would be non-backward-compatible there also. Anyway, defaults can effectively be set in templates/modules by the users.

Reply:
> We definitely need static defaults in the API

+1. New defaults ("Always populated with the value that it would have if I created a new automobile now without specifying the number of wheels") are definitely fine- and if they weren't, that'd be 100% a client problem.

In the cases we've ran into, there's been a static value for new resources after the default was introduced internally (~O(weeks - months prior to release)) and a different value returned for older resources (typically returning null due to missing backfills, which ~O(months) later gets corrected 

> The ability to have different "migrated" vs "newly created" defaults could be useful, I suspect - would it cause Terraform issues?

Yeah 🙁 . Defaults are static in Terraform, mostly, and the cases where they aren't aren't easy to use.